### DISCUSSION TITLE
Handling PATH breakage on Windows with Python 3.8+

### DISCUSSION THREAD

**Discussion by anderslanglands** [AcademySoftwareFoundation/rez#1346] [Q&A] [answered]
As of Python 3.8, PATH is no longer used for loading DLLs. https://docs.python.org/3.8/whatsnew/3.8.html#bpo-36085-whatsnew

For packages that use PATH on Windows like LD_LIBRARY_PATH on Linux (all of mine are like this) I’m not really sure what to do to fix this. 

One option would be to write out a sitecustomize.py that adds all the dll paths and  set USER_BASE to point to it. This will probably break in certain circumstances though. 

Another option would just be to static link everything on Windows. 

I really hope I’m just being stupid and overlooking something obvious here!

**Comment by JeanChristopheMorinPerso** [comment, marked as answer]
Libraries on PyPI just call [os.add_dll_directory](https://docs.python.org/3.8/library/os.html#os.add_dll_directory) in their `__init__.py`.

Other options are what you said (including all similar/derivative options) and you can also make sure the DLLs are in the same path as your Python code (if I'm not mistaken).

But since this is more a question than an issue, I'll convert it to a discussion.

**Reply by anderslanglands** [reply]
Right but that means modifying eg usdview to add the DLL paths for USD but also tbb and boost_python (for starters). Which kinda defeats the whole point of rez right?

**Reply by JeanChristopheMorinPerso** [reply]
Rez is language agnostic and should stay like this. It has `rez-pip`, but we plan to move `rez-pip` outside of rez to really make it even clearer that rez is language agnostic (and also for other reasons that I'm not gonna list here).

As for usdview, boost_python, etc, I'd report that problem to these packages. USD already does the trick I talked about with `os.add_dll_directory`, see https://github.com/PixarAnimationStudios/USD/blob/7a5f8c4311fed3ef2271d5e4b51025fb0f513730/pxr/base/tf/__init__.py#L33. So you'd have to see if your version of USD has that and see if usdview might need something similar.

As for boost_python, when you create C/C++ python extensions, I think the DLL (if not statically linked, in which case you would only have a `pyd` file) has to be at the same level as the `pyd` file.

I'm sorry if I don't have more solutions or answers. It is kind of what it is...

**Reply by JeanChristopheMorinPerso** [reply]
I forgot to mention that if rez was to have a special feature that would handle this situation magically for our users, it would cause more troubles that it would fix. That is as soon as your packages would be used outside of rez (say you have to share a plugin with a third-party that doesn't use rez or you have to package it to PyPI (either the real PyPI or an internal behind the firewall solution) or anything like that, your packages that relied on that magic rez behavior would probably break. I think such a situation would be hard to debug for some users.

**Reply by christian-korneck** [reply]
jfwiw: it's probably the worst hack, but you can bring back loading DLLs from `%PATH%` by changing [this line](https://github.com/python/cpython/blob/264b3ddfd561d97204ffb30be6a7d1fb0555e560/Lib/ctypes/__init__.py#L346) in `lib/ctypes/__init__.py`  to `winmode=0x8`. Wish there was an env var for this. (Using this to solve an obscure non-rez problem).

**Reply by anderslanglands** [reply]
You're right, although that's the same situation as the current behviour: if your package relies on PATH/LD_LIBRARY_PATH being set to find required shared libraries then you need to bundle those libraries up somehow if you want to distribute it outside of rez.

Thanks for the pointer on the USD handling. Looks like that was added in v21and should just work. Question is does everyone else handle it the same way? Will need to look...

**Reply by JeanChristopheMorinPerso** [reply]
> Question is does everyone else handle it the same way?

Pretty much yeah. I looked at all the major packages out there that use C and C++ and even fortran and whatever else and they all use something similar. See https://sourcegraph.com/search?q=context:global+os.add_dll_directory+lang:Python+&patternType=literal. You'll see OpenCV, PyTorch, Numpy, etc.

**Reply by anderslanglands** [reply]
I love this. "Good news everybody! We fixed DLL hell by forcing every python package to reimplement the existing OS mechanism in their own way! Nothing can possibly go wrong with this plan..."

Looks like OpenImageIO and OpenVDB don't do it at all. OpenColorIO only does it for their test runner (and then only specifically to add the build directory).

**Reply by JeanChristopheMorinPerso** [reply]
The OCIO Python bindings are built statically, see https://github.com/AcademySoftwareFoundation/OpenColorIO/blob/main/setup.py#L79. OTIO does the same thing. OpenVDB doesn't have a python package (in the sense that they don't distribute it on PyPI) and they only seem to have 4 python files I think?

**Reply by anderslanglands** [reply]
OK but that only works for the PyPI package. The .pyd produced through the cmake build for ocio will still fail unless static linking is forced. So we're back to static linking everything :\

**Reply by JeanChristopheMorinPerso** [reply]
If you are unhappy with a package not handling DLLs correctly, open an issue with them explaining your use case. Point to mainstream packages as examples and they might add some `os.add_dll_directory` calls. That's how I would resolve that situation. They might have explicitly decided to not do that depending on their user base and a bunch of different reasons. For example, in OTIO we know most users simply use the wheels we publish in PyPI and we worked pretty hard to make them work on all platforms and systems from Python 2.7 to Python 3.10.

So that's really a case by case basis, but I still think it's worth discussing it with packages and libraries.

**Reply by anderslanglands** [reply]
Yep you're right. And worst comes to worst I can just patch the offending libraries at build time to do what USD is doing...

**Comment by anderslanglands** [comment]
@christian-korneck good to know that's there as a nuclear option :D I agree I wish there was an environment variable to revert to the old behaviour...

**Reply by anderslanglands** [reply]
Hmm actually @christian-korneck that doesn't seem to work. Just tried patching my python-3.9, then running it and it's still not searching PATH (looking at procmon)

**Reply by JeanChristopheMorinPerso** [reply]
When he posted the link, I looked at the code and had the impression it would only work when you load a DLL from Python. And yeah, that's the case. The `CDLL` class isn't involved when loading DLLs from a `pyd` (since a yd is just a DLL with a custom extension and some other stuff). Or there is something I'm overlooking. Like I don't remember by heart if `pyd`s are loaded using CDLL or not.

**Reply by anderslanglands** [reply]
Yeah definitely doesn't work unfortunately

**Reply by christian-korneck** [reply]
ah yes, my usecase is only with `ctypes`. Haven't tested with `pyd`. Sorry for the confusion.
